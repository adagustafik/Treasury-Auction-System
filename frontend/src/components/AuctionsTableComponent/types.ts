import { BidType } from "../../api/getBidsRequest";
import { FilterType } from "../types";

export interface AuctionTableProp {
  filterOnly: FilterType;
  limit?: number;
  expanded: boolean;
}

export interface BidTypeState {
  bidtype: BidType;
  selected: boolean;
}
