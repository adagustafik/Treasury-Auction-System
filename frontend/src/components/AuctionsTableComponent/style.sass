@use '../../App'

.bid-button
    width: 100%
    font-size: 2vh    