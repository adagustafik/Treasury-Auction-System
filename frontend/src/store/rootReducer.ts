import { combineReducers } from "@reduxjs/toolkit";
import { auctionsReducer } from "./slices/auctions";
import { userReducer } from "./slices/user";
import { bidsReducer } from "./slices/bids";
import { purchasesReducer } from "./slices/purchases";

export const rootReducer = combineReducers({
  userReducer,
  auctionsReducer,
  purchasesReducer,
  bidsReducer,
});
