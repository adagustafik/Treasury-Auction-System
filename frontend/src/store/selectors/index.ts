export * from "./selectUser";
export * from "./selectAuctions";
export * from "./selectBids";
export * from "./selectPurchases";
