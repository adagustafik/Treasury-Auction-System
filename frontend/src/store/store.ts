import {
  configureStore,
  ThunkAction,
  Action,
  AnyAction,
  ThunkDispatch,
} from "@reduxjs/toolkit";
import thunkMiddleware from "redux-thunk";
import { rootReducer } from "./rootReducer";

export const store = configureStore({
  reducer: rootReducer,
  middleware: [thunkMiddleware],
});

export type AppDispatch = typeof store.dispatch &
  ThunkDispatch<RootState, null, AnyAction>;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
