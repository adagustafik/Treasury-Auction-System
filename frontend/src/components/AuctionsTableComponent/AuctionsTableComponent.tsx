import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import toast from "react-hot-toast";
import { Accordion, Button, Form, Modal, Table } from "react-bootstrap";
import { Auction } from "../../api/getAuctionsRequest";
import {
  compareCloseTimesToDate,
  dateConversionToString,
  filterCurrent,
  filterHistory,
  filterRunning,
} from "../../utilities";
import { useAppDispatch, useAppSelector } from "../../store/hooks";
import { selectAuctions, selectUser } from "../../store/selectors";
import { AuctionTableProp, BidTypeState } from "./types";
import "./style.sass";
import { PurchasesTableComponent } from "../PurchasesTableComponent";
import { fetchNewBid } from "../../api/newBidRequest";
import { fetchBidsData } from "../../store/slices/bids";

export function AuctionsTableComponent({
  filterOnly,
  expanded,
  limit = 100,
}: AuctionTableProp) {
  const { auctions } = useAppSelector(selectAuctions);

  let auctionsTop: Auction[] = [];

  if (auctions.length > 0) {
    const auctionsCurrent = auctions.filter((auction: Auction) =>
      filterOnly === "history" ? filterHistory(auction) : filterCurrent(auction)
    );
    auctionsTop = auctionsCurrent.slice(0, limit);
  }

  const { admin } = useAppSelector(selectUser);
  const navigate = useNavigate();
  const dispatch = useAppDispatch();

  const [show, setShow] = useState(false);
  const [validated, setValidated] = useState(false);
  const [bidType, setBidType] = useState<BidTypeState>({
    bidtype: "noncomp",
    selected: false,
  });
  const [amount, setAmount] = useState(0);
  const [rate, setRate] = useState(0);
  const [auctionId, setAuctionId] = useState(0);

  const handleBidShow = () => {
    setShow(true);
  };

  const handleBidHide = () => {
    setShow(false);
  };

  function validationOutsideOfForm(): boolean {
    if (bidType.selected === false || amount <= 0) {
      toast("Please, fill out all required fields.");
      return false;
    }
    if (bidType.bidtype === "comp" && rate <= 0) {
      toast("Rete is required for competitive bids");
      return false;
    }
    return true;
  }

  const handleBidSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    setValidated(true);
    if (validationOutsideOfForm()) {
      const date = new Date();
      const body = {
        time: date.toISOString(),
        type: bidType.bidtype,
        amount,
        rate,
        auctionid: auctionId,
      };
      fetchNewBid(body)
        .then((res) => {
          if (res.success) {
            handleBidHide();
            dispatch(fetchBidsData);
            navigate("/bids");
            toast(res.success);
          }
        })
        .catch((err) => {
          toast(err.message);
        });
    }
  };

  return (
    <>
      {auctionsTop.length > 0 && (
        <Table striped bordered hover>
          <thead>
            <tr>
              <th>Id</th>
              <th>Auction date</th>
              <th>Closing time</th>
              <th>Issue amount</th>
              <th>Issue date</th>
              <th>Maturity date</th>
              <th>Terms</th>
              <th>Yield</th>
              <th>Rate</th>
              <th>Discount</th>
            </tr>
          </thead>
          <tbody>
            {auctionsTop.map((auction: Auction) => (
              <React.Fragment key={auction.auctionId}>
                <tr>
                  <td>{auction.auctionId}</td>
                  <td>{dateConversionToString(auction.auctionDate)}</td>
                  <td>{compareCloseTimesToDate(auction)}</td>
                  <td>{auction.securityOffered}</td>
                  <td>{dateConversionToString(auction.issueDate)}</td>
                  <td>{dateConversionToString(auction.maturityDate)}</td>
                  <td>{auction.termsAndConditions}</td>
                  <td>{auction.yield}</td>
                  <td>{auction.rate}</td>
                  <td>{auction.discount}</td>
                </tr>
                {expanded === true &&
                  admin === false &&
                  filterRunning(auction) && (
                    <tr>
                      <td colSpan={11}>
                        <Button className="bid-button" onClick={handleBidShow}>
                          Bid now
                        </Button>
                        <Modal
                          size="lg"
                          show={show}
                          onHide={handleBidHide}
                          backdrop="static"
                          centered
                        >
                          <Modal.Header closeButton>
                            <Modal.Title>
                              Place your bid before the auction closing time
                            </Modal.Title>
                          </Modal.Header>
                          <Modal.Body>
                            <Form
                              validated={validated}
                              onSubmit={handleBidSubmit}
                            >
                              <h5>SELECT YOUR BID TYPE</h5>
                              <Form.Check
                                required
                                inline
                                type="radio"
                                name="bidType"
                                label="non-competitive"
                                value="noncomp"
                                onChange={() =>
                                  setBidType({
                                    bidtype: "noncomp",
                                    selected: true,
                                  })
                                }
                              />
                              <Form.Check
                                required
                                inline
                                type="radio"
                                name="bidType"
                                label="competitive"
                                value="comp"
                                onChange={() =>
                                  setBidType({
                                    bidtype: "comp",
                                    selected: true,
                                  })
                                }
                              />
                              <h5>BID AMOUNT</h5>
                              <Form.Control
                                required
                                type="number"
                                min="1"
                                value={amount}
                                onChange={(
                                  event: React.ChangeEvent<HTMLInputElement>
                                ) => {
                                  setAmount(Number(event.target.value));
                                  setAuctionId(auction.auctionId);
                                }}
                              />
                              <Form.Control.Feedback type="invalid">
                                Bidding amount is required
                              </Form.Control.Feedback>
                              {bidType.bidtype === "comp" && (
                                <>
                                  <h5>BID RATE</h5>
                                  <Form.Control
                                    required
                                    type="number"
                                    min={auction.rate}
                                    value={rate}
                                    onChange={(
                                      event: React.ChangeEvent<HTMLInputElement>
                                    ) => setRate(Number(event.target.value))}
                                  />
                                  <Form.Control.Feedback type="invalid">
                                    Bidding rate is required at minimum of
                                    Auction rate
                                  </Form.Control.Feedback>
                                </>
                              )}
                              <Button type="submit" className="bid-button">
                                BID NOW
                              </Button>
                            </Form>
                          </Modal.Body>
                        </Modal>
                      </td>
                    </tr>
                  )}

                {expanded === true && auction.resolved === true && (
                  <tr>
                    <td colSpan={11}>
                      <Accordion>
                        <Accordion.Header>Auction Resolution</Accordion.Header>
                        <Accordion.Body>
                          <PurchasesTableComponent
                            filterBy="auction"
                            auctionId={auction.auctionId}
                          />
                        </Accordion.Body>
                      </Accordion>
                    </td>
                  </tr>
                )}
              </React.Fragment>
            ))}
          </tbody>
        </Table>
      )}
      {auctionsTop.length === 0 && (
        <p>There are currently no auctions to display</p>
      )}
    </>
  );
}
