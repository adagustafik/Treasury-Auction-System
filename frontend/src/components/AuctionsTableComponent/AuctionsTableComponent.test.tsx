import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import { store } from "../../store/store";
import { AuctionsTableComponent } from "./AuctionsTableComponent";

it("renders correctly without bidding option limit 5", () => {
  const registerForm = renderer
    .create(
      <Provider store={store}>
        <BrowserRouter>
          <AuctionsTableComponent
            filterOnly="current"
            expanded={false}
            limit={5}
          />
        </BrowserRouter>
      </Provider>
    )
    .toJSON();
  expect(registerForm).toMatchSnapshot();
});
