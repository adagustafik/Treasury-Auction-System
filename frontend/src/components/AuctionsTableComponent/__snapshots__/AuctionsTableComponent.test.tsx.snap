// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly without bidding option limit 5 1`] = `
<p>
  There are currently no auctions to display
</p>
`;
